File: app/src/main/java/com/mbg/mbgsupport/demo/java/singleton/DemoSingletonSecond.java
package com.mbg.mbgsupport.demo.java.singleton;

/***
 * java方式懒汉式创建单例的demo
 * created by canwen2003
 */
public class DemoSingletonSecond {
    private static final String TAG="DemoSingletonSecond";
    private static DemoSingletonSecond instance;
    private DemoSingletonSecond(){

    }
    public static DemoSingletonSecond getInstance(){
        if (instance==null){
            instance=new DemoSingletonSecond();
        }
        return instance;
    }

    public void logInfo(String str){
        System.out.println(TAG+":"+str);
    }
}

File: app/src/main/java/com/mbg/mbgsupport/demo/kotlin/singleton/DemoSingletonSecond.kt
package com.mbg.mbgsupport.demo.kotlin.singleton
/***
 * Kotlin方式懒汉式创建单例的demo
 * created by canwen2003
 */

//私有构造器是为了防止外面创建对象实例化
class DemoSingletonSecond private constructor(){

    fun logInfo(str: String){
        println("${TAG}:$str")
    }

    companion object{
        private const val TAG="DemoSingletonSecond"
        private var instance :DemoSingletonSecond?=null
        get() {
            if (field==null){
                field= DemoSingletonSecond();
            }
            return field;
        }

        fun get():DemoSingletonSecond{
            return instance!!;
        }
    }

}